### DISCUSSION TITLE
Edge IE Mode Support

### DISCUSSION THREAD

**Discussion by GCuser99** [GCuser99/SeleniumVBA#10] [General]
At @yaju's assistance getting started in [issue #9](https://github.com/GCuser99/SeleniumVBA/issues/9#issue-1250684544), I have added first-cut functionality to support Edge in IE Mode.

To get started, I recommend reading the notes below first, install the desired driver version, unzip and open the Excel file above, and then go to test_IEMode module.

Below are my notes and plans for this support:

The IEDriverServer in IE Mode is very flaky with many quirks and limitations that need attention. Since use of IE mode is transitional while legacy websites using ActiveX components get ported to new browsers, full support for this driver may never go into production for this project, especially given all of the issues documented here.

The 32-bit and 64-bit drivers each have advantages and disadvantages.

The biggest shortcoming of the 64-bit version is the VERY SLOW Sendkeys. It can take 3-5 secs to render each character! This is a well-documented problem on the web. There are some additional Option settings that will improve this but then other problems occur such as delays in rendering urls with NavigateTo and slower action chain execution. Sendkeys seems to work fine with the 32-bit driver. Anyway, to fix the 64-bit driver SendKeys problem one must do the following:
```
    Set caps = driver.CreateCapabilities
    caps.SetOption "requireWindowFocus", True 'require the browser window to have focus before interacting with elements.
    caps.SetOption "ignoreProtectedModeSettings", True 'ignore the settings of the Internet Explorer Protected Mode.
```
Using the 32-bit driver has some issues too. For an as yet unknown reason, some longer action chains such as a stacked series of DragAndDrops may not fully complete with the 32-bit driver. The implemented work-around is to break the action sequences into smaller sub-chains with nominal system waits executed between each sub-chain (see ActionChain Perform method for details). The 64-bit driver seems not to have this issue, even though it executes slower with the SendKeys fix noted above.

Considering all of the above, **I STRONGLY RECOMMEND USING THE 32-BIT DRIVER**.

There appears to be an intermittent bug where CloseBrowser sometimes hangs. See [here](https://github.com/SeleniumHQ/selenium/issues/10600) for similar report from others. The workaround implemented is to set the receiveTimeout for some limited duration when sending the Quit command to the driver. If it times-out then TaskKill is used to terminate the driver. Not pretty but seems to work ok.

**EDIT: as of IEDriverServer version 4.6, the CloseBrowser issue has been fixed.**

Other known limitations of IE mode driver:

- Wheel-type actions in action chains not supported (no action scrolls)
- Aria methods not supported (GetAriaLabel & GetAriaRole)
- Multi-session functionality not supported - cannot start more than one session for same driver instance
- Multi-driver instances not supported (must check if driver process is running and kill process before starting a new driver instance).
- Shadowroots are not supported
- Headless mode not supported
- Incognito (InPrivate) mode does not function
- SetProfile method does not function
- SetCookies method does not seem to work
- SwitchToWindow method does not bring switched-to window into foreground (but still seems to work)
- SaveScreenshot method does not work properly - produces blank screenshots
- PrintToPDF command not recognized
- Changing the default file download preferences through driver capabilities not possible
- No existing api for getting latest driver release for auto-update
- For mouse coordinate move actions to work properly, must maximize window first
- SetGeolocation and GetGeolocation methods of WebDriver class not supported

Things yet to do:

- How to get rid of "Only secure content is displayed" IE pop up through capabilities settings?

Any suggested ideas, comments, fixes, and use cases are welcome!

**Comment by GCuser99** [comment]
I dropped support for IE mode for now - can come back to it easily if there is demand.... Let me know here if anyone has thoughts about support for this...

**Reply by ssktgm** [reply]
Hello, GCuser99!!

I would like to resume support for IE mode.
Anything that works at this time is fine.
I am waiting for a good reply.

**Comment by GCuser99** [comment]
@ssktgm, let me see what we can do. The newest version of the IEDriverServer has apparently fixed the CloseBrowser issue mentioned in my post above, which was the most important impediment to inclusion. It looks like the other issues are still outstanding though...

If we add support, are you ok using the 32 bit driver? The 64 bit sendkeys problem still looms large, and I was not satisfied with my work-around.

**Reply by ssktgm** [reply]
If the sendkeys problem has not been resolved, then there is no problem with using a 32-bit driver. Thanks for your quick reply.

**Comment by GCuser99** [comment]
ok @ssktgm, IE Mode functionality added, with all of the caveats enumerated above.  The automated Browser/WebDriver version alignment function installs the latest 32-bit IEDriverServer.exe to your downloads folder (by default).

**Reply by ssktgm** [reply]
I have confirmed that the IE mode has been implemented through an update. I'll provide feedback if I find an unexpected issue. Thank you for your support.

**Comment by GCuser99** [comment]
It's been over two years since we put IE Mode functionality back into SeleniumVBA. However, for a variety of reasons which I won't go into here, it is getting harder to support it. Is there anyone out there that still is using it? If not, I'm tempted to drop it again and just move forward with Chrome/Edge/FF...

**Reply by 6DiegoDiego9** [reply]
Hi Mike, I think I've never used it, so for me it would be okay to drop it

**Reply by yaju** [reply]
I think after two years it will be ok.

**Reply by GHRyunosuke** [reply]
never used it, vote for dropping it.

**Comment by dornech** [comment]
I am not maintainer nor active supporter but IMHO as a recent heavy SeleniumVBA user switching from SeleniumBasic dropping is reasonable. Beside the fact that I never used the old IE mode (and switched to Selenium to get rid of using the Microsoft the MSHTML stuff just to be able to freely use other browsers instead old IE for automated browswer stuff) I understand that the additional effort does not pay off especially not in a freeware solution while probably blocking other things.

**Comment by GCuser99** [comment]
@yaju, just so that I am clear, are you saying that you no longer need IE mode, or do you still need it for two more years?

Thanks to all for your feedback!

**Reply by yaju** [reply]
Hi GCuser99, I think no longer need IE mode
